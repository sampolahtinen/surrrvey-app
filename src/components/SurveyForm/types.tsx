import { APISurveyDetailsResponse } from "../../api/types";

export interface SurveyFormProps {
  title: string;
  subtitle: string;
  questions: APISurveyDetailsResponse["questions"];
  onSubmit: (params: any) => any;
  isCompleted: boolean;
  errorMessage?: string | boolean;
}
