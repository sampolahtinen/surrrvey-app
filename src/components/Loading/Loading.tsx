import styled from "styled-components";
import { colors } from "../../styles/colors";

export default styled.span`
  font-size: 16px;
  color: ${colors.grey};
  opacity: 0.8;
`;
