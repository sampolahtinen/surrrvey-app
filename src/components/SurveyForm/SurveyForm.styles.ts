import styled from "styled-components";
import { animated } from "react-spring";
import { colors } from "../../styles/colors";

export const SuccessMessage = styled(animated.div)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${colors.white};
  flex-wrap: wrap;
  span {
    display: block;
    color: ${colors.green};
    font-size: 64px;
    font-weight: 600;
    width: 100%;
    text-align: center;
  }
  a {
    font-size: 20px;
  }
`;

export const Wrapper = styled.div`
  position: relative;
  width: 500px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 2rem;
  z-index: 10;
`;

export const Title = styled.h1`
  margin-bottom: 2.4rem;
`;

export const Subtitle = styled.h2`
  margin-bottom: 6.4rem;
  color: ${colors.grey};
  font-weight: 300;
  opacity: 0.5;
`;
