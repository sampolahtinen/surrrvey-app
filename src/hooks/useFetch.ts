import { useEffect, useState } from "react";

export function useFetch(
  callback: (params?: any) => Promise<any>,
  params?: any
) {
  const [results, setResults] = useState<any>({});
  const [error, setError] = useState(undefined);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await callback(params);
        setResults(response.data);
        setIsLoading(false);
      } catch (error) {
        setIsLoading(false);
        setError(error);
      }
    };
    fetchData();
  }, [params, callback]);
  return [isLoading, results, error];
}
