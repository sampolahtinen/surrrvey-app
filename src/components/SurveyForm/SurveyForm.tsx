import React, { FC, useState } from "react";
import { SurveyFormProps } from "./types";
import { Link } from "react-router-dom";
import { config, useTransition } from "react-spring";
import QuestionCarousel from "../QuestionCarousel";
import { Subtitle, Form, Title, SuccessMessage } from "./SurveyForm.styles";
import { Button } from "../Elements/Button";
import { APISurveyCompletion } from "../../api/types";
import { colors } from "../../styles/colors";

const SurveyForm: FC<SurveyFormProps> = ({
  title,
  subtitle,
  questions,
  onSubmit,
  isCompleted,
  errorMessage
}) => {
  const [currentQuestion, setCurrentQuestion] = useState(0);

  const initialState = {
    completion: []
  };

  const [selectedOptions, setSelectedOptions] = useState<APISurveyCompletion>(
    initialState
  );

  const getQuestionAnswer = (questionId: string) =>
    selectedOptions.completion.find(elem => elem.question_id === questionId)
      ?.value;

  const getQuestionIndex = (questionId: string) => {
    const answers = selectedOptions.completion;
    const questionIndex = answers.findIndex(
      elem => elem.question_id === questionId
    );
    return questionIndex;
  };

  const handleOptionSelection = (question: string) => {
    const currentQuestionId = questions[currentQuestion].id;
    const questionIndex = getQuestionIndex(currentQuestionId);
    // If question has not been answered, append it to the completion array
    if (questionIndex === -1) {
      setSelectedOptions({
        completion: [
          ...selectedOptions.completion,
          {
            question_id: currentQuestionId,
            value: question
          }
        ]
      });
      // if question is found from completion, update its value
    } else {
      const nextState = selectedOptions.completion;
      nextState[questionIndex] = {
        question_id: currentQuestionId,
        value: question
      };
      setSelectedOptions({
        completion: nextState
      });
    }
  };

  const handleSubmit = (e: any) => {
    e.preventDefault();
    onSubmit(selectedOptions);
  };

  const transitions = useTransition(isCompleted, null, {
    from: { opacity: 0, transform: "scaleY(0)", transformOrigin: "top" },
    enter: { opacity: 1, transform: "scaleY(1)", transformOrigin: "top" },
    leave: { opacity: 0, transform: "scaleY(0)", transformOrigin: "bottom" },
    config: {
      ...config.stiff,
      clamp: true
    }
  });

  return (
    <Form>
      <Title>{title}</Title>
      <Subtitle>{subtitle}</Subtitle>
      <QuestionCarousel
        questions={questions}
        onOptionSelect={handleOptionSelection}
        selectedOption={getQuestionAnswer(questions[currentQuestion].id) || ""}
        onQuestionChange={(nextQuestionIndex: number) => {
          setCurrentQuestion(nextQuestionIndex);
        }}
      />
      {currentQuestion === questions.length - 1 && (
        <Button
          type="submit"
          onClick={handleSubmit}
          style={{ marginTop: "2rem", width: "150px", height: "50px" }}
        >
          Submit
        </Button>
      )}
      {transitions.map(
        ({ item, props }) =>
          item && (
            <SuccessMessage style={props}>
              <span>Success!</span>
              <span style={{ fontSize: "20px", color: colors.darkGrey }}>
                Your response was delivered.
              </span>
              <Link to="/surveys">Go back to Survey list</Link>
            </SuccessMessage>
          )
      )}
      {errorMessage && <span>{errorMessage}</span>}
    </Form>
  );
};

export default SurveyForm;
