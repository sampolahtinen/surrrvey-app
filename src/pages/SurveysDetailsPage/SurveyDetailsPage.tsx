import React, { FC } from "react";
import SurveyList from "../../components/SurveyList";
import { useFetch } from "../../hooks/useFetch";
import { getSurveyDetails } from "../../api/surveys";
import Layout from "../../components/Layout/Layout";
import { SurveyDetailsPageProps } from "./types";

const SurveyDetailsPage: FC<SurveyDetailsPageProps> = ({ match }) => {
  const id = match.params.id;
  const [loading, { survey }, error] = useFetch(getSurveyDetails, id);
  console.log(survey);
  return (
    <Layout>
      <div className="surveys-page">
        {/* //TODO: create loading component */}
        {loading && <span>Loading</span>}
        {!loading && (
          <>
            <h1 style={{ textDecoration: "underline" }}>{survey.title}</h1>
            <h2 style={{ textDecoration: "underline" }}>{survey.tagline}</h2>
          </>
        )}
        {!error && <span>{error}</span>}
      </div>
    </Layout>
  );
};

export default SurveyDetailsPage;
