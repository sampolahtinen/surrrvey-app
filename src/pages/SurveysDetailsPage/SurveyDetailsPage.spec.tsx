import React from "react";
import { shallow } from "enzyme";
import SurveyDetailsPage from "./SurveyDetailsPage";
import SurveyForm from "../../components/SurveyForm";
import Loading from "../../components/Loading";

const mockSurveyDetails = {
  survey: {
    id: "001",
    title: "test",
    tagline: "testtag",
    questions: [
      {
        id: "Q868b4b",
        title: "What is the best soccer team in the world?",
        options: [
          "BVB",
          "Bayern Munich",
          "Manchester United",
          "FC Barcelona",
          "Some other team"
        ]
      }
    ]
  }
};

const mockMatch = {
  params: { id: "001" },
  isExact: false,
  path: "",
  url: ""
};

jest.mock("../../hooks/useFetch");
import { useFetch } from "../../hooks/useFetch";

describe("SurveyDetailsPage Component", () => {
  useFetch.mockReturnValue([true, mockSurveyDetails, undefined]);
  const wrapper = shallow(<SurveyDetailsPage match={mockMatch} />);
  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays loading before data has been fetched", () => {
    expect(
      wrapper
        .find(Loading)
        .first()
        .text()
    ).toBe("Loading");
  });
});

describe("SurveyDetailsPage data fetching", () => {
  useFetch.mockReturnValue([false, mockSurveyDetails, undefined]);
  const wrapper = shallow(<SurveyDetailsPage match={mockMatch} />);

  it("displays SurveyForm on successful data fetch", () => {
    expect(wrapper.find(SurveyForm).exists()).toBeTruthy();
    useFetch.mockReset();
  });

  // it("displays error on failed data fetch", () => {
  //   useFetch.mockReset();
  //   useFetch.mockReturnValue([false, mockSurveyDetails, "error"]);
  //   console.log(wrapper.debug());
  //   expect(wrapper.find("span").text()).toBe("error");
  // });
});
