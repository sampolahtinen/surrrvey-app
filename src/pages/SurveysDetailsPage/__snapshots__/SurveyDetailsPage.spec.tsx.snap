// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyDetailsPage Component renders 1`] = `
<Layout>
  <styled.div
    className="survey-deytails-page"
  >
    <span>
      Loading
    </span>
  </styled.div>
</Layout>
`;
